invenio_rdm_records/records/jsonschemas/records/record-v1.0.0.json

```diff
@@ -465,36 +465,46 @@
         },
 
         "locations": {
-          "description": "Geographical locations.",
-          "type": "array",
-          "items": {
-            "type": "object",
-            "additionalProperties": false,
-            "properties": {
-              "geometry": {
-                "type": "object",
-                "properties": {
-                  "type": {
-                    "type": "string"
-                  },
-                  "coordinates": {
-                    "type": "array"
+          "description": "Geographical locations, as a GeoJSON FeatureCollection, with additional constraints.",
+          "$comment": "Everything here must be a FeatureCollection, but those features can have additional root-level properties (foreign members, in GeoJSON speak; see RFC7946 §6.1.",
+          "allOf": [
+            {
+              "$ref": "https://geojson.org/schema/FeatureCollection.json"
```

Comment:
moderate: IIRC a conversation with @slint, we avoided having this sort of references due to requests. This will make a request on each validation? It might result in failures when an instance is proxied or cut from the external world with firewalls, etc.

Reply:
Exactly, `jsonschema` [will use the `requests` library](https://python-jsonschema.readthedocs.io/en/latest/references/#jsonschema.RefResolver.resolve_remote) to fetch remote schemas, and cache them. We mitigated this in two ways:

- Not using remote schemas :see_no_evil:
- Using [custom ref resolvers](https://github.com/inveniosoftware/invenio-opendefinition/blob/9744f8e9b35bcc059ce6489763e944e5be906a75/invenio_opendefinition/resolvers.py#L23-L36), so that we control how/if any requests get done (e.g. then we could "vendor" the JSON Schema in the package and always use a local copy instead of making a remote request). Keep in mind that this is a bit tricky to implement and test, so I'm not sure if it's the best option at the moment

Reply:
I definitely agree that we should avoid external lookups. jsonschema lots you supply a [custom ref resolver](https://python-jsonschema.readthedocs.io/en/stable/references/), which means we can a) supply it locally, and b) forbid external resolution.

I did look at this, and say that we Invenio already uses `jsonresolver`, another Invenio project. invenio-records pays heed to an `invenio_records.jsonresolver` entry point group for hooking stuff in, but there are no instances of it being used. I can play though, and it shouldn't be too difficult to get working.

Reply:
I think here we have to replicate parts of the JSONSchema instead. Would be nicer to just reference GeoJSON schema, but apart from the ref resolvers somehow seeming to work quite magical, then there's other issues:

1) Required properties: When we deposit a draft, we need to store in the same data structure, but we cannot enforce the required properties because then you cannot save a half finished upload.
2) ``properties`` is included, which we don't store at all.
3) There's extra stuff like bbox or geometry collection that we don't know how to deal with, so my preference is for having being more restrictive than what the GeoJSON standard allows.

Reply:
We're now using a copy of the Geometry schema which I've embedded in the schema definitions section. The schema for locations is now tied down to not have properties.